runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4129,10 +4133,10 @@ J9::Z::TreeEvaluator::ardbarEvaluator(TR::Node *node, TR::CodeGenerator *cg)
    // handle the side effects. Then we delegate the evaluation of the remaining
    // children and the load/store operation to the appropriate load/store evaluator.
    TR::Node *sideEffectNode = node->getFirstChild();
-   TR::Register *sideEffectRegister = cg->evaluate(sideEffectNode);
-
+   
    if (cg->comp()->getOption(TR_EnableFieldWatch))
       {
+      TR::Register *sideEffectRegister = cg->evaluate(sideEffectNode);
```

Comment:
Same concern as before. This is definitely a bug. We conditionally evaluate the `sideEffectNode` here but unconditionally decrement its reference count later on. We should always be evaluating the `sideEffectNode`.

Reply:
Right. 
If the sideEffectNode is not evaluated unconditionally, then when evaluating the sideEffectNode it’s children will be correctly decremented. Otherwise if conditionally decremented, we’ll be decrementing the count of the node itself but it’s children will not be decremented and not have an accurate reference count because they may have not been evaluated.

Reply:
It's also the evaluator's responsibility for evaluating its children. The side-effect node can be common'ed in other places and so if this is the first time we see it, then we need to evaluate it here as well.

I'm not sure why this conversation was resolved as I still think this is a bug. Would you be able to explain why we don't need a change here?

Reply:
Per my previous comment, it was in agreement and the change was reverted.